### DISCUSSION TITLE
Proposal: `take`, `skip` LINQ query clauses

### DISCUSSION THREAD

**Discussion by orthoxerox** [dotnet/csharplang#592] [General]
Similar to #94, but covers more methods.

A query expression with a `take` clause
```c#
from x in e
take e2
…
```
is translated into
```c#
from x in ( e ) . Take ( e2 )
…
```

A query expression with a `skip` clause
```c#
from x in e
skip e2
…
```
is translated into
```c#
from x in ( e ) . Skip ( e2 )
…
```

A query expression with a `take while` clause
```c#
from x in e
take while f
…
```
is translated into
```c#
from x in ( e ) . TakeWhile ( x => f )
…
```

A query expression with a `skip while` clause
```c#
from x in e
skip while f
…
```
is translated into
```c#
from x in ( e ) . SkipWhile ( x => f )
…
```

A query expression with a `take until` clause
```c#
from x in e
take until f
…
```
is translated into
```c#
from x in ( e ) . TakeWhile ( x => ! ( f ) )
…
```

A query expression with a `skip until` clause
```c#
from x in e
skip until f
…
```
is translated into
```c#
from x in ( e ) . SkipWhile ( x => ! ( f ) )
…
```

**Comment by HaloFour** [comment]
VB.NET already has some of these LINQ query clauses.  Specifically `Take`, `Take While`, `Skip` and `Skip While`.  Interestingly VB.NET does not have `Take Until` or `Skip Until` despite already having `Until` as a keyword for `Do` loops.  I'd assume that adding it to C# might depend on whether or not `until` was also added as a contextual keyword for loops.

Even if a general purpose continuable clause is added to the language I still think it's worth adding specific clauses to handle these really common cases.

**Comment by tuespetre** [comment]
Loving this proposal _except_ I have to say that I would prefer `take while` and `skip while` over `take until` and `skip until`. I think consistency is important, plus it wouldn't need a new keyword.

**Comment by orthoxerox** [comment]
Prototype being developed in https://github.com/orthoxerox/roslyn/tree/features/linq-take-skip

**Comment by scalablecory** [comment]
See also https://github.com/dotnet/roslyn/issues/100

**Comment by lcsondes** [comment]
`take` and `skip` in particular have widespread support in SQL (through LIMIT and depending on the dialect, OFFSET) and are very commonly used, e.g., for pagination in backend code.
